### PR TITLE
Define the `QiskitDevice.batch_execute` method

### LINKED ISSUE #138: Support running Qiskit jobs with multiple circuits
Qiskit supports to submit multiple circuits combined to one job. All circuits in the job are computed as a batch,  which reduces the waiting time in the queue. This can be used for computing gradients as well as a batch of inputs for a circuit, for eample when using a KerasLayer.

I would like to call the qnode with a list of inputs that which result in one job send to the qiskit device. The same can be done for computing gradients. In my oppinion, this would also require to change the behavior of the KerasLayer to provide a batch computation method.

### PR DESCRIPTION
**Pull request by antalszava**
Changes the `QiskitDevice` class to handle the execution of multiple quantum circuits. This addition uses the batch execution pipeline from PennyLane, and the support to run multiple circuit objects from Qiskit.

Related issue: closes #138

### CODE REVIEW

**Comment by antalszava** [comment]
[sc-7289]

File: pennylane_qiskit/qiskit_device.py [hunk 1]
```diff
@@ -349,3 +377,39 @@ def analytic_probability(self, wires=None):
 
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def batch_execute(self, circuits):
```
Comment (antalszava) [inline pennylane_qiskit/qiskit_device.py:381]: This method inherits its docstring from `QubitDevice`.

File: pennylane_qiskit/qiskit_device.py [hunk 2]
```diff
@@ -326,15 +343,26 @@ def _get_state(self, result):
         # reverse qubit order to match PennyLane convention
         return state.reshape([2] * self.num_wires).T.flatten()
 
-    def generate_samples(self):
+    def generate_samples(self, circuit=None):
```
Comment (antalszava) [inline pennylane_qiskit/qiskit_device.py:346]: This method used to inherit its docstring from `QubitDevice` (now takes the `circuit` kwarg).

File: tests/test_ibmq.py [hunk 3]
```diff
@@ -148,13 +152,14 @@ def test_account_error(monkeypatch):
...
 def test_simple_circuit(token, tol, shots):
+    """Test executing a simple circuit submitted to IBMQ."""
     IBMQ.enable_account(token)
     dev = IBMQDevice(wires=2, backend="ibmq_qasm_simulator", shots=shots)
 
     @qml.qnode(dev)
     def circuit(theta, phi):
         qml.RX(theta, wires=0)
-        qml.RX(phi, wires=0)
+        qml.RX(phi, wires=1)
```
Comment (antalszava) [inline tests/test_ibmq.py:162]: A bug we had.

File: tests/test_integration.py [hunk 4]
```diff
@@ -75,7 +80,7 @@ def circuit(x, y, z):
 
     @pytest.mark.parametrize("d", pldevices)
     @pytest.mark.parametrize("shots", [8192])
-    def test_one_qubit_circuit(self, shots, d, backend, tol):
+    def test_basis_state_and_rot(self, shots, d, backend, tol):
```
Comment (antalszava) [inline tests/test_integration.py:83]: A bug we had (has the same name as the previous test case).

File: tests/test_qiskit_device.py [hunk 5]
```diff
@@ -98,3 +99,82 @@ def test_backend_options_cleaned(self):
...
+
+        dev.reset()
+        tape2_expected = dev.execute(self.tape2)
+
+        assert len(res) == 2
+        assert np.allclose(
+            res[0], tape1_expected, atol=0
+        )
+
+        dev.reset() # We're calling device methods directly, need to reset
```
Comment (rmoyard) [inline tests/test_qiskit_device.py:161]: Not necessary here?

File: tests/test_integration.py [hunk 6]
```diff
@@ -479,3 +484,82 @@ def circuit():
...
+        res = qml.grad(circuit)(x,y)
+        expected = np.array([[-np.sin(y) * np.sin(x), np.cos(y) * np.cos(x)]])
+        assert np.allclose(res, expected, **tol)
+
+        # Check that QiskitDevice.batch_execute was called once
+        assert spy1.call_count == 1
+
+        # Check that run was called twice: for the partial derivatives and for
+        # running the circuit
+        assert spy2.call_count == 2
```
Comment (rmoyard) [inline tests/test_integration.py:565]: Nice!

File: tests/test_integration.py [hunk 7]
```diff
@@ -479,3 +484,82 @@ def circuit():
...
+        if (d[0] == "qiskit.aer" and "aer" not in backend) \
+          or (d[0] == "qiskit.basicaer" and "aer" in backend):
+            pytest.skip("Only the AerSimulator is supported on AerDevice")
+
+        if backend not in state_backends and shots is None:
+            pytest.skip("Hardware simulators do not support analytic mode")
+
+        dev = qml.device(d[0], wires=3, backend=backend, shots=shots)
+
+        spy1 = mocker.spy(QiskitDevice, "batch_execute")
```
Comment (rmoyard) [inline tests/test_integration.py:543]: Does it use `batch_execute` because of the parameter shift diff method?

File: tests/test_ibmq.py [hunk 8]
```diff
@@ -166,6 +171,70 @@ def circuit(theta, phi):
     assert np.allclose(res, expected, **tol)
 
 
+@pytest.mark.parametrize("shots", [1000])
+def test_simple_circuit_with_batch_params(token, tol, shots, mocker):
+    """Test that executing a simple circuit with batched parameters is
+    submitted to IBMQ once."""
+    IBMQ.enable_account(token)
+    dev = IBMQDevice(wires=2, backend="ibmq_qasm_simulator", shots=shots)
```
Comment (rmoyard) [inline tests/test_ibmq.py:179]: Should we add a hardware backend?

File: pennylane_qiskit/qiskit_device.py [hunk 9]
```diff
@@ -304,20 +318,23 @@ def run(self, qcirc):
...
             self._state = self._get_state(result)
 
-    def _get_state(self, result):
+    def _get_state(self, result, experiment=None):
         """Returns the statevector for state simulator backends.
 
         Args:
             result (qiskit.Result): result object
 
+        Keyword Args:
```
Comment (rmoyard) [inline pennylane_qiskit/qiskit_device.py:327]:
```suggestion
```

File: pennylane_qiskit/qiskit_device.py [hunk 10]
```diff
@@ -326,15 +343,26 @@ def _get_state(self, result):
...
         return state.reshape([2] * self.num_wires).T.flatten()
 
-    def generate_samples(self):
+    def generate_samples(self, circuit=None):
+        r"""Returns the computational basis samples generated for all wires.
+
+        Note that PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
+        :math:`q_0` is the most significant bit.
+
+        Keyword Args:
```
Comment (rmoyard) [inline pennylane_qiskit/qiskit_device.py:352]:
```suggestion
        Args:
```

**Review by rmoyard — COMMENTED** [review]
Thank you @antalszava, the PR looks great! I've left some minor comments 💯

File: tests/test_qiskit_device.py (hunk 5 above)
Comment (antalszava) [reply tests/test_qiskit_device.py]: Good catch! Removed.

File: tests/test_integration.py (hunk 7 above)
Comment (antalszava) [reply tests/test_integration.py:543]: Yep! The parameter-shift computation goes through the batch execution pipeline. The exact number of executions will actually change slightly with the QNode change.

File: tests/test_ibmq.py (hunk 8 above)
Comment (antalszava) [reply tests/test_ibmq.py:179]: I don't think that's necessary. Here we test that the `IBMQDevice` works well, the rest should be up to IBMQ (i.e., if it works on a remote simulator, it should work on hardware too, unless we have different logic internally).

**Review by rmoyard — APPROVED** [review]
Thanks @antalszava, it looks good to me 💯

**Pull request merged**